backend/drm/drm.c

```diff
@@ -0,0 +1,597 @@
+	eglTerminate(renderer->egl.disp);
+	eglReleaseThread();
+	eglMakeCurrent(EGL_NO_DISPLAY, EGL_NO_SURFACE, EGL_NO_SURFACE, EGL_NO_CONTEXT);
+
+	gbm_device_destroy(renderer->gbm);
+}
+
+static int find_id(const void *item, const void *cmp_to)
+{
+	const struct wlr_drm_display *disp = item;
+	const uint32_t *id = cmp_to;
+
+	if (disp->connector < *id)
+		return -1;
+	else if (disp->connector > *id)
+		return 1;
+	else
+		return 0;
+}
+
+void wlr_drm_scan_connectors(struct wlr_drm_backend *backend)
+{
+	drmModeRes *res = drmModeGetResources(backend->fd);
+	if (!res) {
+		wlr_log(L_ERROR, "Failed to get DRM resources");
+		return;
+	}
+
+	for (int i = 0; i < res->count_connectors; ++i) {
+		uint32_t id = res->connectors[i];
+
+		drmModeConnector *conn = drmModeGetConnector(backend->fd, id);
+		if (!conn) {
+			wlr_log(L_ERROR, "Failed to get DRM connector");
+			continue;
+		}
+
+		struct wlr_drm_display *disp;
+		int index = list_seq_find(backend->displays, find_id, &id);
+
+		if (index == -1) {
+			disp = calloc(1, sizeof *disp);
+			if (!disp) {
+				wlr_log(L_ERROR, "Allocation failed: %s", strerror(errno));
+				drmModeFreeConnector(conn);
+				continue;
+			}
+
+			disp->renderer = &backend->renderer;
+			disp->state = DRM_DISP_DISCONNECTED;
+			disp->connector = res->connectors[i];
+			snprintf(disp->name, sizeof disp->name, "%s-%"PRIu32,
+				 conn_name[conn->connector_type],
+				 conn->connector_type_id);
+
+			list_add(backend->displays, disp);
+			wlr_log(L_INFO, "Found display '%s'", disp->name);
+		} else {
+			disp = backend->displays->items[index];
+		}
+
+		if (disp->state == DRM_DISP_DISCONNECTED &&
+		    conn->connection == DRM_MODE_CONNECTED) {
+			disp->state = DRM_DISP_NEEDS_MODESET;
+			wlr_log(L_INFO, "Sending modesetting signal for '%s'", disp->name);
+			wl_signal_emit(&backend->signals.display_add, disp);
+
+		} else if (disp->state == DRM_DISP_CONNECTED &&
+		    conn->connection != DRM_MODE_CONNECTED) {
+			disp->state = DRM_DISP_DISCONNECTED;
+			wlr_drm_display_free(disp, false);
+			wlr_log(L_INFO, "Sending destruction signal for '%s'", disp->name);
+			wl_signal_emit(&backend->signals.display_rem, disp);
+		}
+
+		drmModeFreeConnector(conn);
+	}
+
+	drmModeFreeResources(res);
+}
+
+static void free_fb(struct gbm_bo *bo, void *data)
+{
+	uint32_t *id = data;
+
+	if (id && *id)
+		drmModeRmFB(gbm_bo_get_fd(bo), *id);
+
+	free(id);
+}
+
+static uint32_t get_fb_for_bo(int fd, struct gbm_bo *bo)
+{
+	uint32_t *id = gbm_bo_get_user_data(bo);
+
+	if (id)
+		return *id;
+
+	id = calloc(1, sizeof *id);
+	if (!id) {
+		wlr_log(L_ERROR, "Allocation failed: %s", strerror(errno));
+		return 0;
+	}
+
+	drmModeAddFB(fd, gbm_bo_get_width(bo), gbm_bo_get_height(bo), 24, 32,
+		     gbm_bo_get_stride(bo), gbm_bo_get_handle(bo).u32, id);
+
+	gbm_bo_set_user_data(bo, id, free_fb);
+
+	return *id;
+}
+
+static bool display_init_renderer(struct wlr_drm_renderer *renderer,
+		struct wlr_drm_display *disp)
```

Comment:
Wait is display here an output, not a wl_display? Let's use the output terminology strictly to avoid confusion.